[skyl3-r] iP

skyl3-r opened the pull request:
# Barcus
> "Before you eat the elephant, make sure you know what parts you want to eat." - Todd Stocker [(source)](https://www.lifestarr.com/blog/11-to-do-list-quotes-to-help-you-begin-tackling-things-today)

Barcus keeps track of what tasks you want to do (or parts of an elephant you want to eat 🐘 😋 ). It's
- text-based
- easy to use
- ~~silly goofy~~ very serious

All you need to do is,
1. download it from [here](https://github.com/skyl3-r/ip/releases/tag/v0.1)
2. run it
3. add in your tasks
4. let barcus fix your life 👍 

> [!NOTE]
> Ensure you have **Java 17** installed

Features:
- [x] Add tasks
- [x] Delete tasks
- [ ] GUI (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

skyl3-r committed:
Rename Chatbot to Barcus

skyl3-r committed:
Add support for echo

skyl3-r committed:
Add support for list

skyl3-r committed:
Implement Task class

skyl3-r committed:
Add support for marking as donw

skyl3-r committed:
Add different task types which inherit from Task

skyl3-r committed:
Implement support for todo, event and deadline

skyl3-r committed:
Edit files for automated text ui testing

skyl3-r committed:
Add class for specific exception

skyl3-r committed:
Add support for error handling

skyl3-r committed:
Replace array with arraylist to store tasks

skyl3-r committed:
Add support for deleting tasks from list

skyl3-r committed:
Update Task class and subclasses for saving

skyl3-r committed:
Add support for saving tasks

skyl3-r committed:
Merge branch 'branch-Level-7'

skyl3-r committed:
Add support for date reading

skyl3-r committed:
Merge branch 'branch-Level-8'

skyl3-r committed:
Create new files for more OOP

skyl3-r committed:
Add functionality for Storage class

skyl3-r committed:
Add functionality for TaskList class

skyl3-r committed:
Add support for Ui and Parser classes

skyl3-r committed:
Merge branch 'branch-A-MoreOOP'

skyl3-r committed:
Reorganise classes into packages

skyl3-r committed:
Merge branch 'add-gradle-support'

skyl3-r committed:
Add support for gradle

skyl3-r committed:
Add support for JUnit test of Parser and Storage

skyl3-r committed:
Change basename for creation of JAR file

skyl3-r committed:
Add javadocs for first half of classes

skyl3-r committed:
Fix coding standard for files

skyl3-r committed:
Add javadocs for other half

skyl3-r committed:
Add support for Find command

skyl3-r committed:
Merge branch 'branch-A-JavaDoc'

skyl3-r committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/barcus/command/DeleteCommand.java

skyl3-r committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/barcus/task/Task.java

valhrd commented on src/main/java/barcus/command/MarkCommand.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        if (pos > 0 && pos <= tasks.getLength()) {
+//                        tasks.get(pos - 1).markDone();
```
Is this still necessary?

valhrd commented on src/main/java/barcus/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) {
+        if (pos > 0 && pos <= tasks.getLength()) {
+//                        tasks.get(pos - 1).unmarkDone();
```
Is this commented out code still necessary?

valhrd commented on src/main/java/barcus/task/Event.java:
```diff
@@ -0,0 +1,63 @@
+    protected LocalDateTime from;
+    protected LocalDateTime to;
+
+    private final static DateTimeFormatter fromFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
+    private final static DateTimeFormatter toFormatter = DateTimeFormatter.ofPattern("MMM dd yyyy HH:mm");
```
This code looks to be repeated in the Deadline class as well, could we try to "shift" it up to its parent class and handle formatting there instead?

valhrd commented on src/main/java/barcus/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+     * @param toFind string to find
+     * @return true if description contains the substring
+     */
+    public boolean containsSubstring(String toFind) {
```
Just a minor preference of changing this method name to "find" but otherwise there's no issue here.

valhrd commented on src/main/java/barcus/tasklist/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
+    private int length;
```
Is the private field "length" necessary since we can call this.tasks.size() in its place?

valhrd commented on src/main/java/barcus/tasklist/TaskList.java:
```diff
@@ -0,0 +1,119 @@
+ * Class to encapsulate the list of tasks
+ */
+public class TaskList {
+    private ArrayList<Task> tasks;
```
Just a minor preference but maybe rename to taskList or listOfTasks to be a bit more descriptive, otherwise no issues at all.

valhrd commented on src/main/java/barcus/command/Command.java:
```diff
@@ -0,0 +1,14 @@
+ */
+public abstract class Command {
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws BarcusException;
+    public abstract boolean isExit();
```
Many commands override this method to return false except for the ExitCommand class. Is it possible to make this method in the Command class concrete and override it to return true only for the ExitCommand class?
```suggestion
    public boolean isExit() {
        return false;
    }
```

valhrd reviewed (commented):
Nice implementation of the Command class! Only some minor preferences when it comes to naming some fields and simplifying some stuff to avoid repeated code blocks, but overall very well-structured code.

skyl3-r committed:
Change code to adhere to coding standards

skyl3-r committed:
Merge branch 'branch-A-CheckStyle'

skyl3-r committed:
Add support for JavaFX

skyl3-r committed:
Merge branch 'branch-Level-10'

skyl3-r committed:
Edit to solve cannot snapshot error

skyl3-r committed:
Clean up so that adhere to style guide

skyl3-r committed:
Merge branch 'master' into branch-A-Varargs

skyl3-r committed:
Add support for varArgs in mark, unmark and delete

skyl3-r committed:
Merge branch 'branch-A-Varargs'

EyuGongYi commented on src/main/java/barcus/parser/Parser.java:
```diff
@@ -0,0 +1,123 @@
+                        String.join(" ", Arrays.copyOfRange(words, 1, byI)),
+                        String.join(" ", Arrays.copyOfRange(words, byI + 1, words.length)));
+            }
+        } else if (words[0].equals("event")) {
+            List<String> wordsList = Arrays.asList(words);
+            if (!wordsList.contains("/from") || !wordsList.contains("/to")) {
+                throw new BarcusException("please include '/from' and '/to' "
+                        + "as well as dates after each of those words");
+            } else {
+                int fromI = wordsList.indexOf("/from");
+                int toI = wordsList.indexOf("/to");
+                return new AddEventCommand(
+                        String.join(" ", Arrays.copyOfRange(words, 1, fromI)),
+                        String.join(" ", Arrays.copyOfRange(words, fromI + 1, toI)),
+                        String.join(" ", Arrays.copyOfRange(words, toI + 1, words.length)));
+            }
+        } else if (words[0].equals("delete")) {
+            if (words.length == 1) {
+                throw new BarcusException("please have an integer after 'delete'");
+            } else {
+                try {
+                    int[] pos = new int[words.length - 1];
+                    for (int i = 1; i < words.length; i++) {
+                        pos[i - 1] = Integer.parseInt(words[i]);
+                    }
+                    return new DeleteCommand(pos);
+                } catch (NumberFormatException e) {
+                    throw new BarcusException("please have an integer after 'delete'");
+                }
+            }
+        } else if (words[0].equals("find")) {
+            if (words.length == 1) {
+                throw new BarcusException("please include what word(s) you want to find after 'find'");
+            } else {
+                return new FindCommand(String.join(" ", Arrays.copyOfRange(words, 1, words.length)));
+            }
+        }
+
+        return new UnknownCommand();
+    }
```
Could use enum and switch statements to make the code look cleaner? It looks abit harder to read for now.
Do you think you can switch up the logic for checking if the command is valid? for example taking out the words.length check to check for multiple commands at once instead of 1 by 1

EyuGongYi commented on src/main/java/barcus/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+ * Task with a date to do by
+ */
+public class Deadline extends Task {
+    private static final DateTimeFormatter fromFormatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
```
Maybe Deadline and Event can extend of a TimedTask? so that there is no need for multiple formatters.

EyuGongYi commented on src/main/java/barcus/ui/Ui.java:
```diff
@@ -0,0 +1,59 @@
+     * Prints welcome message
+     */
+    public void showWelcome() {
+        talk("Beep bop! Hello I am Barcus, ready to be of assistance!\n"
```
Like how you stored all the constant information into UI and appending to the string as needed depending on the command.

EyuGongYi commented on src/main/java/barcus/storage/Storage.java:
```diff
@@ -0,0 +1,104 @@
+
+        TaskList tasks = new TaskList();
+        try {
+            Scanner s = new Scanner(file);
+            while (s.hasNextLine()) {
+                //String line = s.next();
+                String[] lineSplit = s.nextLine().split(" \\| ");
+
+                if (lineSplit[0].equals("T")) {
+                    tasks.addTask(new Todo(lineSplit[2], lineSplit[1].equals("1")));
+                    //System.out.println("added");
+                } else if (lineSplit[0].equals("D")) {
+                    tasks.addTask(new Deadline(
+                            lineSplit[2],
+                            lineSplit[1].equals("1"),
+                            lineSplit[3]
+                    ));
+                } else if (lineSplit[0].equals("E")) {
+                    tasks.addTask(new Event(
+                            lineSplit[2],
+                            lineSplit[1].equals("1"),
+                            lineSplit[3],
+                            lineSplit[4]
```
Could separate this into a separate Parser class? Can help keep the code clean if you plan to store more types of data other than the current 3

EyuGongYi reviewed (commented):
Good job overall, the code standard is great and i cant really find any major things, just small things to nitpick about

skyl3-r committed:
Add line to ensure that assertion is enabled

skyl3-r committed:
Improve code readability of Parser

Parse method of Parser class was a long method and had deep nesting.

This makes it difficult for the reader to understand the logic.

Let's break the method into smaller methods and unnest some logic.

skyl3-r committed:
Use assert to ensure output and queues are not empty

In Command class, output could be empty before it is returned.

This would lead to an empty String being displayed by the Barcus bot.

Let's add an assert statement to ensure that the command has been
executed before getString runs.

skyl3-r committed:
Merge pull request #1 from skyl3-r/branch-A-Assertions

Add assertion statements

skyl3-r committed:
Merge branch 'master' into branch-A-CodeQuality

skyl3-r committed:
Merge pull request #2 from skyl3-r/branch-A-CodeQuality

Improve code readability of Parser

skyl3-r committed:
Set up Continuous Integration

skyl3-r committed:
Edit mild code style violations

skyl3-r committed:
Add support for adding tags and saving them

skyl3-r committed:
Add functionality for tagging

skyl3-r committed:
Add functionality for untagging tasks

skyl3-r committed:
Add functionality for find to work with tags

skyl3-r committed:
Merge branch 'branch-C-Tagging'

skyl3-r committed:
Add support for welcome and goodbye message to show

skyl3-r committed:
Improve GUI to make different users and commands obvious

skyl3-r committed:
Merge branch 'branch-A-BetterGui'

skyl3-r committed:
Add app title and icon

skyl3-r committed:
Add one JUnit Test for TagCommand execute

skyl3-r committed:
Remove all mentions of UI

skyl3-r committed:
Add User Guide through readme

skyl3-r committed:
Edit emojis and alert markdown of readme

skyl3-r committed:
Add Javadocs to all command classes

skyl3-r committed:
Fix bug involving deleting duplicates

skyl3-r committed:
Fix minor bugs

skyl3-r committed:
Fix fxml files

skyl3-r committed:
Fix minor style issues
